text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,401 @@
+- Start Date: 2018-08-08
+- RFC PR: (leave this empty)
+- Tracking Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Support JavaScript class-based inheritance in `wasm-bindgen` by leveraging
+Rust's trait system.
+
+Inheritance between JavaScript classes is instead turned into inheritance
+between Rust traits.
+
+This idea has been used very successfully in [`stdweb`](https://crates.io/crates/stdweb),
+though the implementation is very different.
+
+# Motivation
+[motivation]: #motivation
+
+Class-based inheritance is very common in JavaScript, especially with the
+[WebIDL bindings](https://heycam.github.io/webidl/#dfn-inherit).
+
+Given how prelevant this pattern is in JavaScript, it would be very useful to
+have a convenient and ergonomic way to make use of this inheritance while
+maintaining optimal code reuse in Rust.
+
+The end goal is that it should be possible to write generic code which can
+work with all sub-classes in the inheritance graph.
+
+As an example, it should be possible to use the
+[`appendChild`](https://developer.mozilla.org/en-US/docs/Web/API/Node/appendChild)
+method with all of the classes which inherit from `Node` (e.g. `HTMLElement`,
+`HTMLDivElement`, `SVGElement`, and many more).
```

Comment:
I agree with your assessment that we should focus on (2).

Reply:
If we accomplish 2, doesn't that necessarily mean that 1 will be supported as well?

Reply:
> I personally think that (1) should be a non goal, since there is no benefit for user code to do `fn takes_element<E: IElement>(elem: E) {...}` over `fn takes_element(elem: Element) {...}` other than potentially saving an `into()` at call sites.

I thought about this some more, and doesn't that statement completely invalidate this RFC? Since you said that writing `<E: IElement>` "only saves an `into()` at call sites", that's also true with the current status quo, since people can always write `html_element.into().append_child(node)`, right?

Reply:
Also, I would like to make an argument in favor of users using generic trait bounds.

I wrote a DOM library called dominator. It contains a [`DomBuilder`](https://github.com/Pauan/rust-dominator/blob/bde76d90a440ed5922cf56b2e2320361c9fa0319/src/dom.rs#L327) struct.

This struct is intentionally designed so that it can work with a wide variety of HTML elements, with correct static typing.

For example, it can work with `Node`, `HTMLElement`, `SVGElement`, `HTMLInputElement`, etc. That is accomplished by having it take a generic `A` type argument.

Then it contains a bunch of methods which are bounded by traits:

[`IEventTarget`](https://github.com/Pauan/rust-dominator/blob/bde76d90a440ed5922cf56b2e2320361c9fa0319/src/dom.rs#L460)
[`INode`](https://github.com/Pauan/rust-dominator/blob/bde76d90a440ed5922cf56b2e2320361c9fa0319/src/dom.rs#L472)
[`IElement`](https://github.com/Pauan/rust-dominator/blob/bde76d90a440ed5922cf56b2e2320361c9fa0319/src/dom.rs#L498)
[`IHtmlElement`](https://github.com/Pauan/rust-dominator/blob/bde76d90a440ed5922cf56b2e2320361c9fa0319/src/dom.rs#L640)

The end result is that users can do things like this:

```rust
html!("input" => HtmlInputElement, {
    .attribute("foo", "bar")
    .style("qux", "corge")
})
```

And that macro will be expanded into this:

```rust
{
    let x: DomBuilder<HtmlInputElement> = DomBuilder::new(create_element_ns("input", HTML_NAMESPACE));
    x.attribute("foo", "bar")
     .style("qux", "corge")
     .into_dom()
}
```

Most of the details are unimportant, but the important thing is that you can use the `html!` macro with a wide variety of DOM types, and then call methods (such as `attribute` or `style`) on those DOM types, and everything will be fully statically typed: you will get a compile-error if you call the wrong method on the wrong DOM type.

The generic bounds like `<A: IElement>` are needed to allow the same method to be called on multiple different DOM types. And code bloat isn't a problem because I've carefully made sure that all of the methods are small and inlined.

I think perhaps you are underestimating how useful generic code is, even for external stuff like the DOM.

Reply:
Regarding `html_element.into().append_child(...)`, this won't work because the rust compiler doesn't know what you are trying to convert into. This is always a downside with generic return values. You would have to expand it as
```rust
let node: Node = html_element.into();
node.append_child(...)
```

As for generic arguments, I agree with you that they are useful and far more ergonomic, but they also fit into a bigger picture of function overloading (along with optional arguments and union type arguments). While I would probably have gone down a path much more similarly to what you are proposing, I think there is a general desire to keep the bindings pretty low level, and so far that has meant avoiding generics. This is definitely something that I want to revisit in the future in the form of a unified function overloading proposal, but I don't think that it should be a blocking concern at the moment.